enhancements/machine-api/machine-health-check-pause-on-upgrade.md

```diff
@@ -0,0 +1,188 @@
+
+Nothing we are aware of today
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Should admins be able to enable / disable this feature? (IMHO no, I don't see a usecase when this has disadvantages)
+
+2. Today there is no way to coordinate multiple users of the pause feature of machineHealthChecks. The value of the
+annotation is supposed to be empty. This means that the annotation might already be set by someone else when the new
+controller want to set it as well. That introduces the question of what to do in this case after cluster upgrade: remove
+the annotation, or keep it? In the latter case: how to keep track whether the controller set the annotation or not (e.g.
+in another annotation?)
+
+### Test Plan
+
+TODO (need to check existing e2e tests, and if we extend them with a test case for this, and how much extra load to CI
+it would add)
+
+### Graduation Criteria
+
+TODO I have no good idea yet
+- if this is alpha, beta or stable (do we need this at all, it doesn't introduce a new API)?
+- if this is Dev Preview, Tech Preview or GA?
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+The new controller lives in the machine-api-operator image, so the upgrades will be driven by the CVO which will fetch
+the right image version as usual.
+The controller does not offer any API or configuration. The controller can be replaced with an older or newer
+version without side effects or risks.
+
+### Version Skew Strategy
+
+This controller has some expectation to other APIs, which are the clusterVersion status conditions, and the
+machineHealthCheck pause annotation. In case any of those changes their API (new condition name or annotation key), the
+controller will stop to work. This will be caught in CI.
+
+## Implementation History
+
+## Drawbacks
+
+The controller can't stop ongoing external remediation, which is executed by an external remediation controller. For
+this to happen we would either need to introduce and implement a pause flag on the external remediation CRD, or the
+external remediation controller needs to look for the pause annotation on the machineHealthCheck resource.
+
+## Alternatives
+
+- Not having this feature at all: cluster admins could pause machineHealthChecks manually, but that is unneeded work,
+  and too easy to forget.
+- Implement this in the machine-healthcheck-controller: this would be an obvious choice on a first look. But since the
+  clusterVersion resource is Openshift specific, it would further diverge Openshift machine API and upstream Cluster
+  API. The goal is to align them over time.
+- Track the status of nodes more fine-grained and continue to remediate nodes that currently are not upgrading: this
+  would be a perfect solution and might be investigated more deeply in the future. For now, with respect to timeline, we
+  want to concentrate on the simple solution of pausing MHCs completely.
```

Comment:
The drawback of doing this on a node-by-node basis is that the remediation is going to be another source of requests to drain nodes, and there may not be capacity to do that. We would therefore have 2 controllers potentially blocking each other from taking action and the cluster wouldn't be upgradeable or fixable.